addon_imps/storage/dataverse.py

```diff
@@ -0,0 +1,235 @@
+from __future__ import annotations
+
+import asyncio
+import re
+from dataclasses import (
+    dataclass,
+    fields,
+)
+
+from addon_toolkit.interfaces import storage
+from addon_toolkit.interfaces.storage import (
+    ItemResult,
+    ItemSampleResult,
+    ItemType,
+)
+
+
+DATAVERSE_REGEX = re.compile(r"^dataverse/(?P<id>\d*)$")
+DATASET_REGEX = re.compile(r"^dataset/(?P<id>\d*)$")
+FILE_REGEX = re.compile(r"^file/(?P<id>\d*)$")
+
+
+@dataclass
+class DataverseStorageImp(storage.StorageAddonHttpRequestorImp):
+    """storage on dataverse
+
+    see https://guides.dataverse.org/en/latest/api/native-api.html
+    """
+
+    async def get_external_account_id(self, _: dict[str, str]) -> str:
+        return ""
+
+    async def list_root_items(self, page_cursor: str = "") -> storage.ItemSampleResult:
+        async with self.network.GET(
+            "mydata/retrieve",
+            query=[
+                ["page", page_cursor],
+                *[("role_ids", role) for role in range(1, 9)],
```

Comment:
dataverse docs say roles are configurable per instance... do we have a sense whether they actually vary in practice, or is it safe to hard-code this?

Reply:
these are all of the standart roles which are present in all dataverses